### PR TITLE
Support Hybrid Composition on Android

### PR DESCRIPTION
**Pull request by bparrishMines**
Fixes https://github.com/flutter/flutter/issues/78578

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format])
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[plugin_tool format]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 1]
```diff
@@ -463,7 +475,43 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
+          controller.addOnPlatformViewCreatedListener(
+            onPlatformViewCreated,
+          );
+
+          controller.create();
+          return controller;
+        },
+      );
+    } else if (defaultTargetPlatform == TargetPlatform.android &&
+        !useAndroidViewSurface) {
```
Comment (cyanglaz) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:514]: nits: this `!useAndroidViewSurface` is unnecessary.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 2]
```diff
@@ -463,7 +475,43 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
       'circlesToAdd': serializeCircleSet(circles),
       'tileOverlaysToAdd': serializeTileOverlaySet(tileOverlays),
     };
-    if (defaultTargetPlatform == TargetPlatform.android) {
+    if (defaultTargetPlatform == TargetPlatform.android &&
+        useAndroidViewSurface) {
+      return PlatformViewLink(
```
Comment (cyanglaz) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:480]: this snippet of creating the platform view looks good to me, i'll still defer to @blasten to confirm.

**Review by cyanglaz — APPROVED** [review]
LGTM % nits. Adding @blasten as he is more familiar with android hybrid composition.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 3]
```diff
@@ -463,7 +475,43 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
+                const <Factory<OneSequenceGestureRecognizer>>{},
+            hitTestBehavior: PlatformViewHitTestBehavior.opaque,
+          );
+        },
+        onCreatePlatformView: (PlatformViewCreationParams params) {
+          final SurfaceAndroidViewController controller =
+              PlatformViewsService.initSurfaceAndroidView(
+            id: params.id,
+            viewType: 'plugins.flutter.io/google_maps',
+            layoutDirection: TextDirection.rtl,
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:498]: do you know why it needs RTL?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 4]
```diff
@@ -463,7 +475,43 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
+            creationParams: creationParams,
+            creationParamsCodec: const StandardMessageCodec(),
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:500]: it's missing `onFocus`. I have to fix this in the other plugins as well. It needs:
```dart
onFocus: () {
  params.onFocusChanged(true);
},
```

File: packages/google_maps_flutter/google_maps_flutter/test/google_map_test.dart [hunk 5]
```diff
@@ -587,4 +588,34 @@ void main() {
...
+            initialCameraPosition: CameraPosition(target: LatLng(10.0, 15.0)),
+          ),
+        ),
+      );
+
+      expect(find.byType(AndroidView), findsOneWidget);
+    },
+  );
+
+  // TODO(bparrishMines): Uncomment once https://github.com/flutter/plugins/pull/4017 has landed.
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter/test/google_map_test.dart:608]: Why is it required in this order?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 6]
```diff
@@ -441,6 +442,17 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
     return channel(mapId).invokeMethod<Uint8List>('map#takeSnapshot');
   }
 
+  /// Set [GoogleMapsFlutterPlatform] to uses [AndroidViewSurface] to build the Google Maps widget.
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:445]:
```suggestion
  /// Set [GoogleMapsFlutterPlatform] to use [AndroidViewSurface] to build the Google Maps widget.
```

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart [hunk 7]
```diff
@@ -441,6 +442,17 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
...
+  ///
+  /// This implementation uses hybrid composition to render the Google Maps
+  /// Widget on Android. This comes at the cost of some performance on Android
+  /// versions below 10. See
+  /// https://github.com/flutter/flutter/wiki/Hybrid-Composition for more
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:450]:
```suggestion
  /// https://flutter.dev/docs/development/platform-integration/platform-views#performance for more
```

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart (hunk 4 above)
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:501]: done

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart (hunk 3 above)
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart]: I misinterpreted https://github.com/flutter/plugins/blob/master/packages/webview_flutter/lib/webview_flutter.dart#L108. 

This value is required for `AndroidViewSurface`, but not for `AndroidView` or `UIKitView`. I'll have to fix this when in `google_maps_flutter`.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart (hunk 1 above)
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart]: done

File: packages/google_maps_flutter/google_maps_flutter/test/google_map_test.dart (hunk 5 above)
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter/test/google_map_test.dart:607]: The tests for the widget are in `google_maps_flutter`, but the changes to the code are in `google_maps_flutter_platform_interface`. However, `google_maps_flutter` pulls the latest version of the platform interface from pub.

**Comment by blasten** [comment]
Thanks Maurice. LGTM.

I'm not sure why the interface deals with widgets.  It's like an interface that is also the implementation?

**Review by blasten — APPROVED** [review]

**Comment by blasten** [comment]
@bparrishMines it looks like the web plugin also needs to be updated. I will ask to see why these plugins are set up this way.

**Comment by blasten** [comment]
@bparrishMines can it be checked it in?

**Comment by bparrishMines** [comment]
@blasten the `ci.yaml` still hasn't ran. I'm also unfamiliar with this test. I'll pull in the latest changes and see if it passes.

**Comment by blasten** [comment]
uh, checks look bad. Maybe retrigger by pushing a new empty commit?

**Comment by bparrishMines** [comment]
@stuartmorgan Have you seen a failure like the above. It looks like its failing when setting the Flutter channel: https://github.com/flutter/plugins/blob/master/.cirrus.yml#L18

**Comment by stuartmorgan** [comment]
Resyncing will fix those failures, but I expect it to pick up various Android failures that started at the same time unfortunately.

**Pull request merged**